src/System.Private.CoreLib/src/System/Threading/Thread.CoreRT.cs

```diff
@@ -165,18 +176,30 @@ public bool IsBackground
             }
             set
             {
-                if (IsDead())
+                ThreadState state = GetThreadState();
+                bool isDead = (state & (ThreadState.Stopped | ThreadState.Aborted)) != 0;
+                if (isDead)
                 {
                     throw new ThreadStateException(SR.ThreadState_Dead_State);
                 }
-                // TODO: Keep a counter of running foregroung threads so we can wait on process exit
+                bool started = (state & ThreadState.Unstarted) == 0;
+                // we changing foreground count only for started threads
+                // on thread start we count it state in `StartThread`
                 if (value)
                 {
-                    SetThreadStateBit(ThreadState.Background);
+                    bool wasBackground = ((SetThreadStateBit(ThreadState.Background) & (int)ThreadState.Background) != 0);
```

Comment:
Is there a risk of the thread starting by the time we get here that would cause the accounting to be off?

Reply:
Exactly =\
There is a problem with logic order in `StartThread`: first we `IncrementRunningForeground(thread)` and then `thread.ClearThreadStateBit(ThreadState.Unstarted)`.
Creator thread itself can't break things since it spinwaits until unstarted bit cleared but it can pass reference to created thread `A` to some thread `B` before calling `A.Start()`. Thus following order of execution is possible:
1) A in `StartThread` -> `IncrementRunningForeground` observed `IsBackground == true` and incremented `s_foregroundRunningCount`
2) B in `IsBackground::set` observed that `ThreadState.Unstarted` is still set and set `ThreadState.Background` without decrementing `s_foregroundRunningCount`

I guess moving `IncrementRunningForeground(thread)` after `thread.ClearThreadStateBit(ThreadState.Unstarted)` in `StartThread` would solve this problem.

Reply:
... swapping call order in `StartThread` would not help:
Thread `B` could read state right before thread `A` had its chance to unset unstarted bit, then `A` will increment foreground count and only then `B` will set background state but will not decrement count since we are not rereading `_threadState`. So we need to call `HasStarted()` in condition to get fresh value.

Reply:
Does CoreCLR have a solution for this problem?

Reply:
There is `ThreadStoreLockHolder`
https://github.com/dotnet/runtime/blob/671f99b4f00e6c47926b09e3dcf899f154ca5d31/src/coreclr/src/vm/threads.cpp#L4580
I guess it's necessary because of `OtherThreadsComplete` counting a lot of fields. I really believe my final idea is a correct one. Going to mentally evaluate possible interleavings further.

Reply:
Can this be solved by updating the thread state using Interlocked.Compare change, and retry if the threadstate changed?

Reply:
I'm pretty confident it's already solved with current code layout. With `TestConcurrentIsBackgroundProperty` harness I was able to hunt last bit of malicious race in this nonatomic sequence of updates
https://github.com/dotnet/corert/pull/8364/commits/9a22494430da146c8b1c914033f3bf37c00312ba#diff-91a8b7320709c8f4c39fb87edc4bc8eaecd252423e04d2b812fc91d0ae5af555L450

Reply:
Agree. Great work on stress testing this!